File: Python/Algorithms/DFS.py

# Basic graph implementation for reference.
class Graph:
    def __init__(self, size):
        self.adj = [[]] * size
        self.size = size

# Iterative version of DFS.
def dfsIterative(G, startVert):
    visited = [False] * G.size
    s = []

    visited[startVert] = True
    s.append(startVert)

    while s:
        vert = s.pop()
        for adjV in G.adj[vert]:
            if not visited[adjV]:
                visited[adjV] = True
                s.append(adjV)

# Recursive implementation.
def dfsRecursive(G, startVert):
    visited = [False] * G.size
    dfsAux(G, startVert, visited)

# Recursive helper method.
def dfsAux(G, v, visited):
    visited[v] = True
    for adjV in G.adj[v]:
        if not visited[adjV]:
            # Do something before visit
            dfsAux(G, adjV, visited)
            # Do something after visiit

File: Java/Algorithms/DFS.java
package Algorithms;

import java.util.LinkedList;
import java.util.Stack;

public class DFS {

	// Primitive definition of a Graph.
	public static class Graph {
		LinkedList<Integer>[] adj;
		int size;
	}
	
	// Iterative DFS implementation.
	public static void dfsIterative(Graph G, int startVert) {
		
		boolean[] visited = new boolean[G.size];
		Stack<Integer> s = new Stack<>();
		
		visited[startVert] = true;
		s.push(startVert);
		
		while (!s.isEmpty()) {
			int vert = s.pop();
			for(int adjV: G.adj[vert]) {
				if (!visited[adjV]) {
					visited[adjV] = true;
					s.push(adjV);
				}
			}
		}
	}
	
	// Recursive DFS that allows for function calls before and after visits.
	public static void dfsRecursive(Graph G, int startVert) {
		boolean[] visited = new boolean[G.size];
		dfsAux(G, startVert, visited);
	}
	
	private static void dfsAux(Graph G, int v, boolean[] visited) {
		visited[v] = true;
		for(int adjV: G.adj[v]) {
			if(!visited[adjV]) {
				dfsAux(G, adjV, visited);
			}
		}
	}
}
